### PR TITLE
Misc fixes

### PR DESCRIPTION
**Pull request by skirpichev**

### CODE REVIEW

**Comment by cbm755** [comment]
Ok parts of this for #671 LGTM.  I had set this up as nonurgent low hanging fruit for new contributors but oh well.

**Comment by skirpichev** [comment]
Thank you for review, @cbm755.

> I had set this up as nonurgent low hanging fruit for new contributors

Unless I miss something, sympy issue is still there.  Waiting for...

File: diofant/solvers/ode.py [hunk 1]
```diff
@@ -361,6 +361,9 @@ def get_numbered_constants(eq, num=1, start=1, prefix='C'):
         raise ValueError("Expected Expr or iterable but got %s" % eq)
 
     atom_set = set().union(*[i.free_symbols for i in eq])
+    functions_set = set().union(*[i.atoms(Function) for i in eq])
+    if functions_set:
+        atom_set |= {Symbol(str(f.func)) for f in functions_set}
```
[the remark this answers is not in the document]
Comment (skirpichev) [reply diofant/solvers/ode.py]: I'm not sure about canonical representation of the function name, maybe I should use ``f.func__name__`` directly, not the str printer.

**Comment by skirpichev** [comment]
Somewhat unrelated, but I would like to know, @cbm755, what do you think about adding Diofant as an optional backend for octsympy?

Disclaimer: I don't use Octave for a long time, but I think such use case may be helpful for the Diofant project.  And for octsympy users too, as there are a lot of fixed bugs, which still valid for sympy (esp., series/limit-related, assumptions system, polys module and polynomial solvers) and some new stuff was added too.  API will be unstable until 1.0, but I think that I can help you solve any compatibility issues.

I have impression, that you don't use too much new code, created for SymPy after the Diofant was forked out (e.g. new stuff in combinatorics or holonomic module, or new solvers).  

Some preliminary work to support the Diofant as backend I did in https://github.com/skirpichev/octsympy/tree/diofant-experiment.  It was much more difficult than porting Mathics to Diofant.  I think, mostly because you overuse calling ``sympify`` via overloaded ``__call__`` for  the SingletonRegistry class.  Usually, you can just create a specific SymPy class instance, like ``Rational`` or ``Symbol``.

**Comment by cbm755** [comment]
Could do.  Multiple backends is good for exposing front-end bugs.  My main concern would be how much technical debt we would take on (e.g., there are already many conditionals on SymPy version).

For sure, I'd be open to PR that make it easier to swap backends.

Briefly, what is the argument *against* constructions like `S(10)/3`?

**Comment by skirpichev** [comment]
On Sat, Aug 11, 2018 at 05:59:42PM -0700, Colin B. Macdonald wrote:
>    Could do. Multiple backends is good for exposing front-end bugs. My main
>    concern would be how much technical debt we would take on (e.g., there are
>    already many conditionals on SymPy version).

I don't know for sure.  Obvivously, new backend would make sense if
it's maintenance burden will be reduced to minimum.

Another my concern, besides S() usage, is using physics module for things
like TensorProduct.  I think, there is the tensor module, isn't?

>    Briefly, what is the argument against constructions like S(10)/3?

This is a syntactic sugar for sympify.  Which can be useful while
interactive work (but I think that something like AST transformations
in IPython is much better to do 10/3 -> Rational(10, 3) substitutions),
but not in the library code.  Python is a dynamic language by
itself and by using S(10) you take away even more type safety.  If
your context expects only integers, why write your code as it would
assume anything.  Explicit is better than implicit.

**Comment by cbm755** [comment]
Re: `S`, that makes sense.  I'd merge a PR that removes (or reduces) library use of this.  And I'll personally try not to use it in that way.

TensorProduct: this is a pretty basic function, does seem a shame to pull in a quantum physics library just for that.  But OTOH, if its not implemented elsewhere...

**Comment by skirpichev** [comment]
On Sun, Aug 12, 2018 at 11:39:20AM -0700, Colin B. Macdonald wrote:
>    Re: S, that makes sense. I'd merge a PR that removes (or reduces) library
>    use of this.

I'll try to make a pr, when collect all such cases.

>    TensorProduct: this is a pretty basic function, does seem a shame to pull
>    in a quantum physics library just for that. But OTOH, if its not
>    implemented elsewhere...

Yes, I think too - that's not the right place.  Maybe tensor module,
if it doesn't have already something similar.

**Pull request merged**